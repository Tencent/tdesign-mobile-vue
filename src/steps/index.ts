import LocalSteps from './steps';
import LocalStepItem from './step-item';
import { withInstall, WithInstallType } from '../shared';
import { TdStepsProps, TdStepItemProps } from './type';

import './style';

export * from './type';
export type StepItemProps = TdStepItemProps;
export type StepsProps = TdStepsProps;

export const StepItem: WithInstallType<typeof LocalStepItem> = withInstall(LocalStepItem);
export const Steps: WithInstallType<typeof LocalSteps> = withInstall(LocalSteps);
