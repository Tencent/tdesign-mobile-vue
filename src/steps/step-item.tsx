import { computed, inject, defineComponent, getCurrentInstance, h, onUnmounted, ComponentInternalInstance } from 'vue';
import { CloseIcon, CheckIcon } from 'tdesign-icons-vue-next';

import props from './step-item-props';
import config from '../config';
import { usePrefixClass } from '@/hooks/useClass';
import { useTNodeJSX } from '@/hooks/tnode';

const { prefix } = config;
const name = `${prefix}-step-item`;

export default defineComponent({
  name,
  props,
  setup(props, context) {
    const stepItemClass = usePrefixClass('step-item');
    const renderTNodeJSX = useTNodeJSX();

    const internalInstance = getCurrentInstance();
    const { proxy } = internalInstance as ComponentInternalInstance;
    const stepsProvide: any = inject('stepsProvide', undefined);
    stepsProvide.relation(proxy);
    const iconDefault = {
      check: h(CheckIcon),
      close: h(CloseIcon),
    };

    const index = computed(() => stepsProvide.state.children.indexOf(proxy));
    const isLastChild = computed(() => {
      return index.value === (stepsProvide.sequence === 'positive' ? stepsProvide.state.children.length - 1 : 0);
    });

    const theme = computed(() => stepsProvide.theme);
    const dot = computed(() => theme.value === 'dot');
    const icon = computed(() => props.icon || context.slots.icon);

    const current = computed(() => stepsProvide.current.value || stepsProvide.defaultCurrent || 0);
    const stepsStatus = computed(() => stepsProvide.currentStatus);
    const readonly = computed(() => stepsProvide.readonly);
    const currentStatus = computed(() => {
      const { status } = props;
      if (status !== 'default') return status;
      if (index.value === current.value) return stepsStatus.value;
      if (index.value < current.value) return 'finish';
      return status;
    });

    const rootClassName = computed(() => [
      stepItemClass.value,
      `${stepItemClass.value}--${stepsProvide.layout}`,
      { [`${stepItemClass.value}--default`]: readonly.value },
      { [`${stepItemClass.value}--${currentStatus.value}`]: currentStatus.value },
    ]);
    const iconWrapperClassName = computed(() => [
      `${stepItemClass.value}__anchor`,
      `${stepItemClass.value}__anchor--${stepsProvide.layout}`,
    ]);
    const dotClass = computed(() => [
      `${stepItemClass.value}__dot`,
      `${stepItemClass.value}__dot--${currentStatus.value}`,
    ]);

    const iconClassName = computed(() => [
      { [`${stepItemClass.value}__icon`]: icon.value },
      { [`${stepItemClass.value}__icon--${currentStatus.value}`]: icon.value },
      { [`${stepItemClass.value}__circle`]: !icon.value },
      { [`${stepItemClass.value}__circle--${currentStatus.value}`]: !icon.value },
    ]);
    const contentClass = computed(() => [
      `${stepItemClass.value}__content`,
      `${stepItemClass.value}__content--${stepsProvide.layout}`,
      {
        [`${stepItemClass.value}__content--${stepsProvide.layout}`]: isLastChild.value,
      },
      {
        [`${stepItemClass.value}__content--last`]: isLastChild.value,
      },
    ]);
    const tilteClass = computed(() => [
      `${stepItemClass.value}__title`,
      `${stepItemClass.value}__title--${currentStatus.value}`,
      `${stepItemClass.value}__title--${stepsProvide.layout}`,
    ]);
    const descriptionClass = computed(() => [
      `${stepItemClass.value}__description`,
      `${stepItemClass.value}__description--${currentStatus.value}`,
      `${stepItemClass.value}__description--${stepsProvide.layout}`,
    ]);
    const extraClass = computed(() => [
      `${stepItemClass.value}__extra`,
      `${stepItemClass.value}__extra--${currentStatus.value}`,
      `${stepItemClass.value}__extra--${stepsProvide.layout}`,
    ]);
    const separatorClass = computed(() => [
      stepItemClass.value,
      `${stepItemClass.value}__line`,
      `${stepItemClass.value}__line--${currentStatus.value}`,
      `${stepItemClass.value}__line--${stepsProvide.sequence}`,
      `${stepItemClass.value}__line--${stepsProvide.layout}`,
      `${stepItemClass.value}__line--${stepsProvide.theme}`,
    ]);
    const onClickIcon = (e: MouseEvent) => {
      if (!readonly.value) {
        stepsProvide.onClickItem(index.value, current.value, e);
      }
    };

    onUnmounted(() => {
      stepsProvide.removeRelation(proxy);
    });
    return () => {
      const renderIconContent = () => {
        if (icon.value) {
          return renderTNodeJSX('icon') || null;
        }
        if (currentStatus.value === 'error') {
          return iconDefault.close;
        }
        if (currentStatus.value === 'finish') {
          return iconDefault.check;
        }
        return index.value + 1;
      };

      return (
        <div class={rootClassName.value} onClick={onClickIcon}>
          <div class={iconWrapperClassName.value}>
            {dot.value ? (
              <div class={dotClass.value}></div>
            ) : (
              <div class={iconClassName.value}>{renderIconContent()}</div>
            )}
          </div>
          <div class={contentClass.value}>
            <div class={tilteClass.value}>
              {renderTNodeJSX('title')}
              {renderTNodeJSX('titleRight')}
            </div>
            <div class={descriptionClass.value}>{renderTNodeJSX('content')}</div>
            <div class={extraClass.value}>{renderTNodeJSX('extra')}</div>
          </div>
          {!isLastChild.value && <div class={separatorClass.value}></div>}
        </div>
      );
    };
  },
});
