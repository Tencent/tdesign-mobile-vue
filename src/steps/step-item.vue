<template>
  <div :class="rootClassName" @click="onClickIcon">
    <!-- icon -->
    <div :class="iconWrapperClassName">
      <div v-if="dot" :class="dotClass"></div>
      <div v-else :class="iconClassName">
        <t-node :content="iconContent"></t-node>
      </div>
    </div>
    <!-- content -->
    <div :class="contentClass">
      <div :class="tilteClass">
        <t-node :content="titleContent"></t-node>
        <t-node :content="titleRightContent" />
      </div>
      <div :class="descriptionClass">
        <t-node :content="descContent"></t-node>
      </div>
      <div :class="extraClass">
        <t-node :content="extraContent"></t-node>
      </div>
    </div>
    <!-- line -->
    <div v-if="!isLastChild" :class="separatorClass"></div>
  </div>
</template>

<script lang="ts">
import { computed, inject, defineComponent, getCurrentInstance, ComponentInternalInstance, h, onUnmounted } from 'vue';
import { CloseIcon, CheckIcon } from 'tdesign-icons-vue-next';

import StepItemProps from './step-item-props';
import { renderTNode, TNode } from '../shared';

import config from '../config';

const { prefix } = config;
const name = `${prefix}-step-item`;

export default defineComponent({
  name,
  components: { TNode },
  props: StepItemProps,
  setup(props, context) {
    const internalInstance = getCurrentInstance();
    const { proxy } = internalInstance as ComponentInternalInstance;
    const stepsProvide: any = inject('stepsProvide', undefined);
    stepsProvide.relation(proxy);
    const iconDefault = {
      check: h(CheckIcon),
      close: h(CloseIcon),
    };

    const index = computed(() => stepsProvide.state.children.indexOf(proxy));
    const isLastChild = computed(() => {
      return index.value === (stepsProvide.sequence === 'positive' ? stepsProvide.state.children.length - 1 : 0);
    });

    const theme = computed(() => stepsProvide.theme);
    const dot = computed(() => theme.value === 'dot');
    const icon = computed(() => props.icon || context.slots.icon);
    const iconNode = computed(() => renderTNode(internalInstance, 'icon'));

    const iconContent = computed(() => {
      if (icon.value) {
        return iconNode.value;
      }
      if (currentStatus.value === 'error') {
        return iconDefault.close;
      }
      if (currentStatus.value === 'finish') {
        return iconDefault.check;
      }
      return index.value + 1;
    });
    const titleContent = computed(() => renderTNode(internalInstance, 'title'));
    const titleRightContent = computed(() => renderTNode(internalInstance, 'titleRight'));
    const descContent = computed(() => renderTNode(internalInstance, 'content'));
    const extraContent = computed(() => renderTNode(internalInstance, 'extra'));

    const current = computed(() => stepsProvide.current.value || stepsProvide.defaultCurrent || 0);
    const stepsStatus = computed(() => stepsProvide.currentStatus);
    const readonly = computed(() => stepsProvide.readonly);
    const currentStatus = computed(() => {
      const { status } = props;
      if (status !== 'default') return status;
      if (index.value === current.value) return stepsStatus.value;
      if (index.value < current.value) return 'finish';
      return status;
    });

    const rootClassName = computed(() => [
      name,
      `${name}--${stepsProvide.layout}`,
      { [`${name}--default`]: readonly.value },
      { [`${name}--${currentStatus.value}`]: currentStatus.value },
    ]);
    const iconWrapperClassName = computed(() => [`${name}__anchor`, `${name}__anchor--${stepsProvide.layout}`]);
    const dotClass = computed(() => [`${name}__dot`, `${name}__dot--${currentStatus.value}`]);

    const iconClassName = computed(() => [
      { [`${name}__icon`]: icon.value },
      { [`${name}__icon--${currentStatus.value}`]: icon.value },
      { [`${name}__circle`]: !icon.value },
      { [`${name}__circle--${currentStatus.value}`]: !icon.value },
    ]);
    const contentClass = computed(() => [
      `${name}__content`,
      `${name}__content--${stepsProvide.layout}`,
      {
        [`${name}__content--${stepsProvide.layout}`]: isLastChild.value,
      },
      {
        [`${name}__content--last`]: isLastChild.value,
      },
    ]);
    const tilteClass = computed(() => [
      `${name}__title`,
      `${name}__title--${currentStatus.value}`,
      `${name}__title--${stepsProvide.layout}`,
    ]);
    const descriptionClass = computed(() => [
      `${name}__description`,
      `${name}__description--${currentStatus.value}`,
      `${name}__description--${stepsProvide.layout}`,
    ]);
    const extraClass = computed(() => [
      `${name}__extra`,
      `${name}__extra--${currentStatus.value}`,
      `${name}__extra--${stepsProvide.layout}`,
    ]);
    const separatorClass = computed(() => [
      name,
      `${name}__line`,
      `${name}__line--${currentStatus.value}`,
      `${name}__line--${stepsProvide.sequence}`,
      `${name}__line--${stepsProvide.layout}`,
      `${name}__line--${stepsProvide.theme}`,
    ]);

    const onClickIcon = (e: MouseEvent) => {
      if (!readonly.value) {
        stepsProvide.onClickItem(index.value, current.value, e);
      }
    };

    onUnmounted(() => {
      stepsProvide.removeRelation(proxy);
    });

    return {
      isLastChild,
      dot,
      name,
      theme,
      index,
      current,
      readonly,
      iconNode,
      stepsStatus,
      descContent,
      iconContent,
      onClickIcon,
      titleContent,
      titleRightContent,
      extraContent,
      rootClassName,
      currentStatus,
      dotClass,
      iconWrapperClassName,
      iconClassName,
      separatorClass,
      contentClass,
      tilteClass,
      descriptionClass,
      extraClass,
    };
  },
});
</script>
