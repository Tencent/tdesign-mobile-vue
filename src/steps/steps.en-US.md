:: BASE_DOC ::

## API
### Steps Props

name | type | default | description | required
-- | -- | -- | -- | --
current | String / Number | - | `v-model` and `v-model:current` is supported | N
defaultCurrent | String / Number | - | uncontrolled property | N
currentStatus | String | process | options：default/process/finish/error | N
layout | String | horizontal | options：horizontal/vertical | N
options | Array | - | Typescript：`Array<TdStepItemProps>` | N
readonly | Boolean | false | \- | N
separator | String | line | options：line/dashed/arrow | N
theme | String | default | options：default/dot | N
onChange | Function |  | Typescript：`(current: string \| number, previous: string \| number, context?: { e?: MouseEvent }) => void`<br/> | N

### Steps Events

name | params | description
-- | -- | --
change | `(current: string \| number, previous: string \| number, context?: { e?: MouseEvent })` | \-

### StepItem Props

name | type | default | description | required
-- | -- | -- | -- | --
content | String / Slot / Function | '' | Typescript：`string \| TNode`。[see more ts definition](https://github.com/Tencent/tdesign-mobile-vue/blob/develop/src/common.ts) | N
icon | Boolean / Slot / Function | true | Typescript：`boolean \| TNode`。[see more ts definition](https://github.com/Tencent/tdesign-mobile-vue/blob/develop/src/common.ts) | N
status | String | default | options：default/process/finish/error。Typescript：`StepStatus` `type StepStatus = 'default' \| 'process' \| 'finish' \| 'error'`。[see more ts definition](https://github.com/Tencent/tdesign-mobile-vue/tree/develop/src/steps/type.ts) | N
subStepItems | Array | [] | Typescript：`SubStepItem[]` `interface SubStepItem { status: StepStatus, title: string }`。[see more ts definition](https://github.com/Tencent/tdesign-mobile-vue/tree/develop/src/steps/type.ts) | N
title | String / Slot / Function | '' | Typescript：`string \| TNode`。[see more ts definition](https://github.com/Tencent/tdesign-mobile-vue/blob/develop/src/common.ts) | N
