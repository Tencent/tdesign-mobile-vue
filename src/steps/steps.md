:: BASE_DOC ::

## API
### Steps Props

| 名称           | 类型            | 默认值     | 说明                                                                                                                                                                                                                                                                                                                              | 必传 |
| -------------- | --------------- | ---------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---- |
| current        | String / Number | -          | 当前步骤，即整个步骤条进度。默认根据步骤下标判断步骤的完成状态，当前步骤为进行中，当前步骤之前的步骤为已完成，当前步骤之后的步骤为未开始。如果每个步骤没有设置 value，current 值为步骤长度则表示所有步骤已完成。如果每个步骤设置了自定义 value，则 current = 'FINISH' 表示所有状态完成。支持语法糖 `v-model` 或 `v-model:current` | N    |
| defaultCurrent | String / Number | -          | 当前步骤，即整个步骤条进度。默认根据步骤下标判断步骤的完成状态，当前步骤为进行中，当前步骤之前的步骤为已完成，当前步骤之后的步骤为未开始。如果每个步骤没有设置 value，current 值为步骤长度则表示所有步骤已完成。如果每个步骤设置了自定义 value，则 current = 'FINISH' 表示所有状态完成。非受控属性                                | N    |
| currentStatus  | String          | process    | 用于控制 current 指向的步骤条的状态。可选项：default/process/finish/error                                                                                                                                                                                                                                                         | N    |
| layout         | String          | horizontal | 步骤条方向，有两种：横向和纵向。可选项：horizontal/vertical                                                                                                                                                                                                                                                                       | N    |
| readonly       | Boolean         | false      | 只读状态                                                                                                                                                                                                                                                                                                                          | N    |
| theme          | String          | default    | 步骤条风格。可选项：default/dot                                                                                                                                                                                                                                                                                                   | N    |
| sequence       | String          | positive   | 步骤条顺序。可选项：positive/reverse                                                                                                                                                                                                                                                                                              | N    |
| onChange       | Function        |            | TS 类型：`(current: string \| number, previous: string \| number, context?: { e?: MouseEvent }) => void`<br/>当前步骤发生变化时触发                                                                                                                                                                                               | N    |

### Steps Events

| 名称   | 参数                                                                                    | 描述                   |
| ------ | --------------------------------------------------------------------------------------- | ---------------------- |
| change | `(current: string \| number, previous: string \| number, context?: { e?: MouseEvent })` | 当前步骤发生变化时触发 |

### StepItem Props

| 名称       | 类型                      | 默认值  | 说明                                                                                                                                                                                                                                                                                            | 必传 |
| ---------- | ------------------------- | ------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---- |
| content    | String / Slot / Function  | ''      | 步骤描述。TS 类型：`string \| TNode`。[通用类型定义](https://github.com/Tencent/tdesign-mobile-vue/blob/develop/src/common.ts)                                                                                                                                                                  | N    |
| icon       | Boolean / Slot / Function | true    | 图标，默认显示内置图标，也可以自定义图标，值为 false 则不显示图标。优先级大于 `status` 定义的图标。TS 类型：`boolean \| TNode`。[通用类型定义](https://github.com/Tencent/tdesign-mobile-vue/blob/develop/src/common.ts)                                                                        | N    |
| status     | String                    | default | 当前步骤的状态：默认状态（未开始）、进行中状态、完成状态、错误状态。可选项：default/process/finish/error。TS 类型：`StepStatus` `type StepStatus = 'default' \| 'process' \| 'finish' \| 'error'`。[详细类型定义](https://github.com/Tencent/tdesign-mobile-vue/tree/develop/src/steps/type.ts) | N    |
| title      | String / Slot / Function  | ''      | 标题。TS 类型：`string \| TNode`。[通用类型定义](https://github.com/Tencent/tdesign-mobile-vue/blob/develop/src/common.ts)                                                                                                                                                                      | N    |
| titleRight | String / Slot / Function  | ''      | 标题右侧数据 仅支持 layout  = 'vertical' 时。TS 类型：`string \| TNode`。[通用类型定义](https://github.com/Tencent/tdesign-mobile-vue/blob/develop/src/common.ts)                                                                                                                               | N    |



### CSS Variables
组件提供了下列 CSS 变量，可用于自定义样式。
| 名称                                    | 默认值                       | 描述 |
| --------------------------------------- | ---------------------------- | ---- |
| --td-step-item-circle-font-size         | 14px                         | -    |
| --td-step-item-circle-size              | 22px                         | -    |
| --td-step-item-default-circle-bg        | @bg-color-secondarycontainer | -    |
| --td-step-item-default-circle-color     | @font-gray-3                 | -    |
| --td-step-item-default-dot-border-color | @component-border            | -    |
| --td-step-item-default-icon-color       | @font-gray-3                 | -    |
| --td-step-item-default-title-color      | @font-gray-3                 | -    |
| --td-step-item-description-color        | @font-gray-3                 | -    |
| --td-step-item-dot-size                 | 8px                          | -    |
| --td-step-item-error-circle-bg          | @error-color-1               | -    |
| --td-step-item-error-circle-color       | @error-color                 | -    |
| --td-step-item-error-dot-border-color   | @error-color                 | -    |
| --td-step-item-error-icon-color         | @error-color                 | -    |
| --td-step-item-error-title-color        | @error-color                 | -    |
| --td-step-item-finish-circle-bg         | @brand-color-light           | -    |
| --td-step-item-finish-circle-color      | @brand-color                 | -    |
| --td-step-item-finish-dot-border-color  | @brand-color                 | -    |
| --td-step-item-finish-icon-color        | @brand-color                 | -    |
| --td-step-item-finish-line-color        | @brand-color                 | -    |
| --td-step-item-finish-title-color       | @font-gray-1                 | -    |
| --td-step-item-line-color               | @component-border            | -    |
| --td-step-item-process-circle-bg        | @brand-color                 | -    |
| --td-step-item-process-circle-color     | @font-white-1                | -    |
| --td-step-item-process-dot-border-color | @brand-color                 | -    |
| --td-step-item-process-icon-color       | @brand-color                 | -    |
| --td-step-item-process-title-color      | @brand-color                 | -    |
