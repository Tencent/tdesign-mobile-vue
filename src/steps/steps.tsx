import { provide, defineComponent, reactive, ComponentInternalInstance, computed } from 'vue';
import props from './props';
import config from '../config';
import { TdStepsProps } from './type';
import { useDefault } from '../shared';
import { useTNodeJSX } from '../hooks/tnode';
import { usePrefixClass } from '../hooks/useClass';

const { prefix } = config;
const name = `${prefix}-steps`;
export default defineComponent({
  name,
  props,
  emits: ['update:current', 'update:modelValue', 'change'],
  setup(props, context) {
    const stepsClass = usePrefixClass('steps');
    const baseClass = computed(() => [
      stepsClass.value,
      `${stepsClass.value}--${props.layout}`,
      `${stepsClass.value}--${props.sequence}`,
      { [`${stepsClass.value}--readonly`]: props.readonly },
    ]);
    const renderTNodeJSX = useTNodeJSX();

    const [current, setCurrent] = useDefault<TdStepsProps['current'], TdStepsProps>(
      props,
      context.emit,
      'current',
      'change',
    );

    interface TState {
      children: ComponentInternalInstance[];
    }

    const state: TState = reactive({
      children: [],
    });

    const relation = (child: ComponentInternalInstance) => {
      child && state.children.push(child);
    };

    const removeRelation = (child: ComponentInternalInstance) => {
      state.children = state.children.filter((item) => item !== child);
    };

    const onClickItem = (cur: TdStepsProps['current'], prev: TdStepsProps['current'], e: MouseEvent) => {
      setCurrent(cur, prev, { e });
    };

    provide('stepsProvide', {
      ...props,
      state,
      current,
      relation,
      removeRelation,
      onClickItem,
    });

    return () => {
      const renderContent = renderTNodeJSX('default') || null;
      return <div class={baseClass.value}>{renderContent}</div>;
    };
  },
});
