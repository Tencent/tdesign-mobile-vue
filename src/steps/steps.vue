<template>
  <div :class="baseClass">
    <slot>
      <t-step-item
        v-for="(item, index) in options"
        :key="index"
        :title="item.title"
        :content="item.content"
        :icon="item.icon"
        :status="item.status"
      ></t-step-item>
    </slot>
  </div>
</template>

<script lang="ts">
import { toRefs, provide, computed, defineComponent, reactive, ComponentInternalInstance } from 'vue';
import StepsProps from './props';
import TStepItem from './step-item.vue';
import config from '../config';
import { TdStepsProps } from './type';
import { useDefault } from '../shared';

const { prefix } = config;
const name = `${prefix}-steps`;
export default defineComponent({
  name,
  components: {
    TStepItem,
  },
  props: StepsProps,
  emits: ['update:current', 'update:modelValue', 'change'],
  setup(props, context) {
    const baseClass = computed(() => [
      name,
      `${name}--${props.layout}`,
      { [`${name}--readonly`]: props.readonly },
      `${name}--${props.theme}-anchor`,
    ]);

    const [current, setCurrent] = useDefault<TdStepsProps['current'], TdStepsProps>(
      props,
      context.emit,
      'current',
      'change',
    );

    const state = reactive({
      children: [] as ComponentInternalInstance[],
    });

    const relation = (child: ComponentInternalInstance) => {
      child && state.children.push(child);
    };

    const onClickItem = (cur: TdStepsProps['current'], prev: TdStepsProps['current'], e: MouseEvent) => {
      setCurrent(cur, prev, { e });
    };

    provide('stepsProvide', {
      ...props,
      state,
      current,
      relation,
      onClickItem,
    });

    return {
      baseClass,
      onClickItem,
      ...toRefs(props),
    };
  },
});
</script>
