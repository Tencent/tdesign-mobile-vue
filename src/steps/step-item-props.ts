/* eslint-disable */

/**
 * 该文件为脚本自动生成文件，请勿随意修改。如需修改请联系 PMC
 * */

import { TdStepItemProps } from '../steps/type';
import { PropType } from 'vue';

export default {
  /** 步骤描述 */
  content: {
    type: [String, Function] as PropType<TdStepItemProps['content']>,
    default: '',
  },
  /** 图标，默认显示内置图标，也可以自定义图标，值为 false 则不显示图标。优先级大于 `status` 定义的图标 */
  icon: {
    type: [Boolean, Function] as PropType<TdStepItemProps['icon']>,
    default: true,
  },
  /** 当前步骤的状态：默认状态（未开始）、进行中状态、完成状态、错误状态 */
  status: {
    type: String as PropType<TdStepItemProps['status']>,
    default: 'default' as TdStepItemProps['status'],
    validator(val: TdStepItemProps['status']): boolean {
      if (!val) return true;
      return ['default', 'process', 'finish', 'error'].includes(val);
    },
  },
  /** 子步骤条，仅支持 layout  = 'vertical' 时 */
  subStepItems: {
    type: Array as PropType<TdStepItemProps['subStepItems']>,
    default: (): TdStepItemProps['subStepItems'] => [],
  },
  /** 标题 */
  title: {
    type: [String, Function] as PropType<TdStepItemProps['title']>,
    default: '',
  },
};
