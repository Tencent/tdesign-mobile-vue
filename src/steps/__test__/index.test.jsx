import { mount } from '@vue/test-utils';
import { describe, it, expect, vi } from 'vitest';
import { nextTick, ref } from 'vue';
import { Steps, StepItem } from '../../steps';
const items = [
  {
    title: '步骤描述1',
    content: '辅助信息文字最多两行',
  },
  {
    title: '步骤描述1',
    content: '辅助信息文字最多两行',
  },
];
const itemStatus = [
  {
    title: '步骤描述1',
    content: '辅助信息文字最多两行',
    status: 'error',
  },
  {
    title: '步骤描述1',
    content: '辅助信息文字最多两行',
    status: 'process',
  },
];
describe('steps', () => {
  describe('props', () => {
    it(': theme', async () => {
      const current = ref(0);
      const theme = 'dot';
      const layout = 'vertical';
      const onChange = vi.fn();
      const event = {};
      const wrapper = mount({
        setup() {
          return () => (
            <Steps v-model={current.value} theme={theme} layout={layout} onChange={onChange}>
              {{
                default: items.map((item, index) => {
                  return <StepItem title={item.title} content={item.content} />;
                }),
              }}
            </Steps>
          );
        },
      });
      const $stepItems = wrapper.findAllComponents(StepItem);
      expect($stepItems.length).toEqual(items.length);
      $stepItems.map(async (item, index) => {
        expect(item.find(`.t-step-item__dot`).text()).toEqual('');
        expect(item.find(`.t-step-item__title`).text()).toEqual(items[index].title);
        expect(item.find(`.t-step-item__description`).text()).toEqual(items[index].content);
      });
    });

    it(': readonly', async () => {
      const current = ref(0);
      const onChange = vi.fn();
      const event = {};
      const wrapper = mount({
        setup() {
          return () => (
            <Steps v-model={current.value} readonly={true} onChange={onChange}>
              {{
                default: items.map((item, index) => {
                  return <StepItem title={item.title} content={item.content} />;
                }),
              }}
            </Steps>
          );
        },
      });
      const $stepItems = wrapper.findAllComponents(StepItem);
      expect($stepItems.length).toEqual(items.length);
      $stepItems.map(async (item, index) => {
        expect(item.find(`.t-step-item__circle`).text()).toEqual(String(index + 1));
        expect(item.find(`.t-step-item__title`).text()).toEqual(items[index].title);
        expect(item.find(`.t-step-item__description`).text()).toEqual(items[index].content);
        await item.find(`.t-step-item`).trigger('click', event);
      });

      // readonly:true, 只读模式, 不会触发 change
      expect(onChange).toHaveBeenCalledTimes(0);
    });

    it(': layout', () => {
      const current = ref(1);
      const layout = 'vertical';
      const wrapper = mount({
        setup() {
          return () => (
            <Steps v-model={current.value} readonly={true} layout={layout}>
              {{
                default: itemStatus.map((item, index) => {
                  return <StepItem title={item.title} content={item.content} status={item.status} />;
                }),
              }}
            </Steps>
          );
        },
      });
      expect(wrapper.element).toMatchSnapshot();
      const $stepItems = wrapper.findAllComponents(StepItem);
      expect($stepItems.length).toEqual(itemStatus.length);

      $stepItems.map((item, index) => {
        const classes = item.find(`.t-step-item`).classes();
        expect(classes.includes(`t-step-item--${itemStatus[index].status}`)).toBeTruthy();
      });
    });
  });

  describe('StepItem Slots', () => {
    it(': content && title', async () => {
      const current = ref(1);
      const stepItems = ref(items.slice());
      const wrapper = mount({
        setup() {
          return () => (
            <Steps v-model={current.value}>
              {{
                default: stepItems.value.map((item, index) => {
                  return <StepItem title={item.title} content={item.content} />;
                }),
              }}
            </Steps>
          );
        },
      });

      stepItems.value.pop();
      await nextTick();

      const $stepItems = wrapper.findAllComponents(StepItem);
      expect($stepItems.length).toEqual(stepItems.value.length);
      $stepItems.map((item, index) => {
        expect(item.find(`.t-step-item__title`).text()).toEqual(stepItems.value[index].title);
        expect(item.find(`.t-step-item__description`).text()).toEqual(stepItems.value[index].content);
      });
    });
  });

  describe('event', () => {
    it(': change', async () => {
      const current = ref(1);
      const onChange = vi.fn();
      const event = {};
      const wrapper = mount({
        setup() {
          return () => (
            <Steps v-model={current.value} onChange={onChange}>
              {{
                default: items.map((item, index) => {
                  return <StepItem title={item.title} content={item.content} />;
                }),
              }}
            </Steps>
          );
        },
      });
      const $stepItems = wrapper.findAllComponents(StepItem);
      expect($stepItems.length).toEqual(items.length);
      // 点击第 2 项, 触发 change, 此时 current = 2,
      await $stepItems[1].find(`.t-step-item`).trigger('click', event);
      expect(onChange).toHaveBeenCalledTimes(1);

      $stepItems.map(async (item, index) => {
        //  current = 2, 则 2 状态为 process,  2 之前的子项状态为 finish,  2 之后的子项状态为 default
        const classes = item.find(`.t-step-item`).attributes('class');
        if (index < 1) {
          expect(classes.includes(`t-step-item--finish`)).toBeTruthy();
        } else if (index === 1) {
          expect(classes.includes(`t-step-item--process`)).toBeTruthy();
        } else {
          expect(classes.includes(`t-step-item--default`)).toBeTruthy();
        }
      });
    });
  });
});
