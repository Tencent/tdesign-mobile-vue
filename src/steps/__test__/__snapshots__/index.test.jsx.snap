// Vitest Snapshot v1

exports[`steps > props > : layout 1`] = `
<div
  class="t-steps t-steps--vertical t-steps--positive t-steps--readonly"
  modelvalue="1"
>
  
  <div
    class="t-step-item t-step-item--vertical t-step-item--default t-step-item--error"
  >
    <div
      class="t-step-item__anchor t-step-item__anchor--vertical"
    >
      <div
        class="t-step-item__circle t-step-item__circle--error"
      >
        <svg
          class="t-icon t-icon-close"
          fill="none"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M7.05 5.64L12 10.59l4.95-4.95 1.41 1.41L13.41 12l4.95 4.95-1.41 1.41L12 13.41l-4.95 4.95-1.41-1.41L10.59 12 5.64 7.05l1.41-1.41z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="t-step-item__content t-step-item__content--vertical t-step-item__content--vertical t-step-item__content--last"
    >
      <div
        class="t-step-item__title t-step-item__title--error t-step-item__title--vertical"
      >
        步骤描述1
        
      </div>
      <div
        class="t-step-item__description t-step-item__description--error t-step-item__description--vertical"
      >
        辅助信息文字最多两行
      </div>
      <div
        class="t-step-item__extra t-step-item__extra--error t-step-item__extra--vertical"
      >
        <!---->
      </div>
    </div>
    <!---->
  </div>
  <div
    class="t-step-item t-step-item--vertical t-step-item--default t-step-item--process"
  >
    <div
      class="t-step-item__anchor t-step-item__anchor--vertical"
    >
      <div
        class="t-step-item__circle t-step-item__circle--process"
      >
        2
      </div>
    </div>
    <div
      class="t-step-item__content t-step-item__content--vertical t-step-item__content--vertical t-step-item__content--last"
    >
      <div
        class="t-step-item__title t-step-item__title--process t-step-item__title--vertical"
      >
        步骤描述1
        
      </div>
      <div
        class="t-step-item__description t-step-item__description--process t-step-item__description--vertical"
      >
        辅助信息文字最多两行
      </div>
      <div
        class="t-step-item__extra t-step-item__extra--process t-step-item__extra--vertical"
      >
        <!---->
      </div>
    </div>
    <!---->
  </div>
  
</div>
`;
